---
title: "MAGs analysis"
author: "Mam Malick Sy Ndiaye"
date: "2022-12-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)
library(viridis)
library(gtools)
library(circlize)
library(ggbeeswarm)
source("../useful_func.R")

fig_dir<- "../../../Figures/manuscript/raw"
if (!dir.exists(fig_dir)){
  dir.create(fig_dir, recursive = T)
} 
tab_dir<- "../../../Supplementary_tabs"
```

# Metgenomes assembly

```{r open assembly statistics}
asmbl_stats_p<- "../../../results/assembly/HF_assembly/all_HFassemblies_metaspades_summary_tab.txt"
asmbl_stats<- fread(asmbl_stats_p)
```

```{r plot contigs filtering}
asmbl_stats2<- asmbl_stats %>% group_by(sample, accepted) %>%
  dplyr::summarise("bp"= sum(length),
                   "total_contigs"=length(contig)) %>%
  dplyr::mutate(sam_type=ifelse(grepl("P", sample), "virome", "bacteriome"),
                sam_name=substring(sample,1, nchar(sample)-1))

ggplot(asmbl_stats2, aes(x=reorder(sam_name,as.numeric(sam_name)),
                         y=bp,
                         fill=factor(accepted, levels = c("yes", "no"))))+
  geom_bar(stat = "identity", position=position_dodge())+
  scale_fill_discrete( name="Passes Filter?")+
  labs(x="")+
  theme_classic()+
  facet_grid(sam_type~., scales = "free_y")


ggplot(asmbl_stats2, aes(x=reorder(sam_name,as.numeric(sam_name)),
                         y=total_contigs,
                         fill=factor(accepted, levels = c("yes", "no"))))+
  geom_bar(stat = "identity", position=position_dodge())+
  scale_fill_discrete( name="Passes Filter?")+
  labs(x="")+
  theme_classic()+
  facet_grid(sam_type~., scales = "free_y")
```


# MAGs binning & Filtering
```{r open and format checkm data}
checkm_alls_p<- "../../../results/MAG_binning/checkm_QC/summary/all_MAGs_stats.tsv"
checkm_filt_p<- "../../../results/MAG_binning/checkm_QC/summary/filtered_MAGs_stats.tsv"


chk_drep_p<- "../../../results/reference_db/data_tables/Chdb.csv"

# get isolates checkm info
iso_check<- fread(chk_drep_p, header=T, sep=",", check.names = F)  %>%
  filter( !grepl("MAG", `Bin Id`)) %>%
  mutate(sample=NA,
         `Bin Id`=gsub(".fna", "", `Bin Id`))
checkm_alls<- read.csv(checkm_alls_p, header=T, sep="\t", check.names = F) 
checkm_filt<- read.csv(checkm_filt_p, header=T, sep="\t", check.names = F) %>%
  rbind(., iso_check)%>%
  mutate(type=ifelse(grepl("MAG", `Bin Id`), "MAG", "isolate"))

filt_mags<- checkm_filt$`Bin Id`
checkm_alls<- checkm_alls %>%
  dplyr::mutate("good"=ifelse(`Bin Id` %in% filt_mags, "yes", "no"))
```

```{r Plot MAGs filtering}
checkm_alls2<- checkm_alls %>% 
  dplyr::mutate(filtered=ifelse(good=="yes", 1, 0)) %>%
  group_by(sample) %>%
  dplyr::summarise(count_good=sum(filtered),
                   tot_count=length(sample)) %>%
  melt()

n<-as.numeric(nrow(checkm_filt))

sam_levs= unique(paste(sort(as.numeric(gsub("B", "", checkm_alls2$sample))), "B", sep=""))

MAGs_filtering_plt<- ggplot(checkm_alls2,aes(x=factor(sample, levels = sam_levs), y=value, 
                          fill=factor(variable, levels = c("tot_count", "count_good")))) +
  geom_bar(stat="identity", position = position_dodge())+
  # geom_segment(aes(x = 0.5, xend = 49.5, y = -2, yend = -2),
  #                  size=10, col="lightpink2") +
  # geom_segment(aes(x = 49.5, xend = 73.5, y = -2, yend = -2),
  #              size=10, col="lightcyan2") +
  # annotate("text", x = 25, y = -2, label = "Switzerland", face="bold") +
  # annotate("text", x = 61, y = -2, label = "Japan", face="bold") +
  scale_y_continuous(breaks = seq(0,70, by=5))+
  scale_fill_manual(name = "", labels = c("Total MAGs", "Filtered MAGs"), values=c("lightskyblue3", "lightpink"))+
  labs(x="", y= "# MAGs")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, face = "bold"),
        axis.text.y = element_text(face = "bold"),
        legend.position = "top",
        legend.text = element_text(size=15,face="bold")); MAGs_filtering_plt
```

```{r MAGs retained  vs. Discarded}
checkm_alls %>% group_by(sample, good) %>%
  dplyr::summarise(count=length(good)) %>%
  ggplot(.,aes(x=good, y=count)) +
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  labs(x="", y= "number of MAGs")+
  theme_classic()
```
# MAGs taxonomic composition
```{r}
core=c("Bombilactobacillus" , "Commensalibacter", "Lactobacillus","Bifidobacterium","Gilliamella",  "Frischella", "Snodgrassella",  "Bartonella","Apibacter")
genome_info<-  read.csv("../../../results/inStrain/profile_B/summary_data/all_genome_info_filtered.tsv", header=T, sep="\t")
present_ids<- unique(genome_info$id)

clust_info<- fread("../../../results/reference_db_filtered/summary_data_tables/clust_filtered.tsv") %>%
  mutate(type=ifelse(grepl("MAG", genome), "MAG", "isolate"),
         community=ifelse(genus %in% core, "Honeybee Associated", "Other"))

clust_info_present<- clust_info %>%
  filter(id %in% present_ids)
```

From the bacterial samples we obtained 478 high-quality (completeness 75% and contamination <10%) bacterial MAGs (bMAGs) (Fig. 1d,Table S3). These high-quality bMAGs, together with 220 bacterial genomes of isolates from the gut of A. mellifera and other bee species, were clustered at 95% ANI to obtain “species-level” bacterial OTUs (bOTUs). We obtained a total of 92 bOTUs.
```{r}
length(clust_info$`Bin Id`[clust_info$type=="MAG"])
length(clust_info$`Bin Id`[clust_info$type=="isolate"])
length(unique(clust_info$id))
```

We detected 53 bOTU in the 49 individual honeybees. Of these, 37 were from honeybee-associated bacterial genera and contained 86% of the high-quality bMAGs (Table S3, Fig. S2). The rest of the bMAGs were associated with environmental bacteria or bee pathogens.
```{r}
length(unique(clust_info_present$id))
length(unique(clust_info_present$id[clust_info_present$community=="Honeybee Associated"]))

all_MAGs<- length(unique(clust_info$genome))
pres_core<- length(unique(clust_info_present$genome[clust_info_present$community=="Honeybee Associated"]))
pres<- length(unique(clust_info_present$genome))

(pres/all_MAGs)*10
(pres_core/all_MAGs)*100
```

*Fig. S2*
```{r}
MAG_pie_plt<- clust_info_present %>% group_by(genus) %>%
  mutate(n_rel=length(genome)/pres)%>%
  ggplot(., aes(x="", y=n_rel, fill=factor(genus, levels = levs)))+
  geom_bar(stat="identity", width=1) +
  scale_fill_manual(values=cols, name="Genus")+
  theme_classic()+
  theme(
        legend.text = element_text(size=14, face="bold"),
        legend.title=element_text(size=14, face="bold"),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks = element_blank())+
  coord_polar("y", start=0);MAG_pie_plt
```




```{r}

```


# TRASH 
# MAGs taxonomic classification
```{r funtion to parse taxonomic information of GTDB-TK}
get_taxa_cols<- function(entry){
  tax_l<- unlist(strsplit(entry, ";"))
  tax_l_ref<- lapply(tax_l, function(x) gsub(".*__", "", x))
  
  tax_l_ref[tax_l_ref==""]<- NA
  
  names(tax_l_ref)<- c("domain", "phylum", "class", "order", "family", "genus", "species")
  tax_l_ref$species<- gsub("__*.","", tax_l_ref$species)
  
  return(tax_l_ref)
}
```

```{r open GTDB-TK output and format it}
gtdb_p<- "../../../results/MAG_binning/gtdbtk_classification/gtdbtk.bac120.summary.tsv"
gtdb_raw<- read.csv(gtdb_p, header=T, sep = "\t")

gtdb<- gtdb_raw %>% group_by(user_genome) %>%
  dplyr::mutate("domain"= get_taxa_cols(classification)$domain,
                "pylum"= get_taxa_cols(classification)$pylum,
                "class"= get_taxa_cols(classification)$class,
                "order"= get_taxa_cols(classification)$order,
                "family"= get_taxa_cols(classification)$family,
                "genus"= get_taxa_cols(classification)$genus,
                "species"= get_taxa_cols(classification)$species)

core=c("Bombilactobacillus" , "Commensalibacter", "Lactobacillus","Bifidobacterium","Gilliamella",  "Frischella", "Snodgrassella",  "Bartonella","Apibacter")
```

```{r merge checkm and GTDB-TK data}
gtdb_QC<- left_join(checkm_filt, gtdb, by=c("Bin Id"="user_genome")) %>%
  group_by(genus) %>%
  dplyr::mutate("n_genus"= length(genus))

gtdb_core<- gtdb_QC %>% filter(genus %in% core)
```

```{r Plot MAGs taxonomical inf per samples}
levs<- get_gen_col_abs(gtdb_QC$genus)[[1]]
cols<-get_gen_col_abs(gtdb_QC$genus)[[2]]

gtdb_comp<- gtdb_QC %>% group_by(sample, genus) %>%
  dplyr::summarise(n=length(genus)) %>%
  group_by(sample) %>%
  dplyr::mutate(n_rel=n/sum(n),
                type= ifelse(genus %in% get_core(), "core", "non-core"))

gtdb_comp %>% group_by(type) %>%
  reframe(n=length(type))


recovered_MAGs_plt<- ggplot(gtdb_comp, aes(x=factor(sample, levels = sam_levs), y=n, fill=factor(genus, levels = levs)))+
  geom_histogram(stat = "identity", position = "stack")+
  geom_segment(aes(x = 0.5, xend = 49.5, y = -3, yend = -3),
                   size=5, col="lightpink2") +
  geom_segment(aes(x = 49.5, xend = 73.5, y = -3, yend = -3),
               size=5, col="lightcyan2") +
  annotate("text", x = 25, y = -3, label = "Switzerland", face="bold") +
  annotate("text", x = 61, y = -3, label = "Japan", face="bold") +
  scale_fill_manual(values=cols, name="Genus")+
  labs(y="# MAGs", x="")+
  scale_y_continuous(breaks=seq(0,15, by=1))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, face="bold"),
        axis.text.y = element_text(face="bold"),
        legend.text = element_text(size=10,face="bold"),
        legend.title = element_text(size=15, face="bold")); recovered_MAGs_plt

ggplot(gtdb_comp, aes(x=factor(sample, levels = sam_levs), y=n_rel, fill=factor(genus, levels = levs)))+
  geom_histogram(stat = "identity", position = "stack")+
  scale_fill_manual(values=cols, name="Genus")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, face="bold"))

MAG_pie_plt<- gtdb_comp %>% filter(!is.na(genus)) %>%
  ggplot(., aes(x="", y=n_rel, fill=factor(genus, levels = levs)))+
  geom_bar(stat="identity", width=1) +
  scale_fill_manual(values=cols, name="Genus")+
  theme_classic()+
  theme(
        legend.text = element_text(size=14, face="bold"),
        legend.title=element_text(size=14, face="bold"),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks = element_blank())+
  coord_polar("y", start=0);MAG_pie_plt

# ggsave(paste(fig_dir, "MAG_piechart.png", sep=""), MAG_pie_plt,units = "mm",
#        height = 210, width = 297, dpi = 600)
```



```{r Plot MAGs genomics stats}
gtdb_QC$lab<-paste("n=", gtdb_QC$n_genus, sep="")

# Data for average genome size have been taken from
# Bombilactobacillus (i.e., Firm4): Zheng et al., 2020 (Microbiology Society)
# Other genera: Zheng et al., 2019 (PNAS; Dataset_S01) 
avg_size<- c("Bombilactobacillus"=1.82e6, 
             "Lactobacillus"=1.91e6, 
             "Bifidobacterium"=1.98e6, 
             "Gilliamella"=2.71e6, 
             "Frischella"=2.69e6,
             "Snodgrassella"=2.43e6,
             "Bartonella"=2.64e6,
             "Apibacter"=2.5e6,
             "Commensalibacter"=2.01e6)

gtdb_QC$avg_size<- avg_size[gtdb_QC$genus]

mag_genome_size_plt<- ggplot(gtdb_QC, aes(x=genus, y=`Genome size (bp)`))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_point(aes(y=avg_size), shape=95, size=12, col="red")+
  geom_text(aes(genus, 0, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1));mag_genome_size_plt

mag_core_genome_size_plt<- gtdb_QC %>%
  filter(genus %in% get_core()[get_core()!="Bombella"])%>%
  ggplot(aes(x=genus, y=`Genome size (bp)`))+
  coord_flip()+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_point(aes(y=avg_size), shape="|", size=17, col="red")+
  geom_text(aes(genus, 0, label = lab),vjust = 1, check_overlap = TRUE)+
  labs(x="")+
  theme_classic()+
  theme(axis.text.y = element_text(face = "bold"),
        axis.text.x = element_text(face = "bold", size=10));mag_core_genome_size_plt

mag_completeness_plt<- ggplot(gtdb_QC, aes(x=genus, y=Completeness))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_text(aes(genus, 0, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1));mag_completeness_plt

mag_contamination_plt<- ggplot(gtdb_QC, aes(x=genus, y=Contamination))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_text(aes(genus, -1, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1));mag_contamination_plt

mag_SH_plt<- ggplot(gtdb_QC, aes(x=genus, y=`Strain heterogeneity`))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_text(aes(genus, -1, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1));mag_SH_plt
```
```{r}

ggplot(gtdb_core, aes(x=`Genome size (bp)`, y= Completeness, group=genus))+
  geom_point()+
  geom_smooth(method="lm",formula=y~log(x))+
  theme_classic()+
  facet_wrap(.~genus)

ggplot(gtdb_core, aes(x=Completeness, y= `Strain heterogeneity`, group=genus))+
  geom_point()+
  geom_smooth(method="lm")+
  theme_classic()+
  facet_wrap(.~genus)
```


# Clustering Info
```{r}
assoc_df<- fread("../../../data/metadata/Genomes_database_metadata_MN_Kelle_nnred.txt")

assoc<- setNames(assoc_df$Host, assoc_df$Locus_tag)

clust_p<- "../../../results/reference_db_filtered/summary_data_tables/clust_filtered.tsv"
clust_win_p<- "../../../results/reference_db_filtered/summary_data_tables/clust_filtered_winners.tsv"




clust<- read.csv(clust_p, header=T, sep="\t") %>% 
  group_by(species) %>%
  dplyr::mutate(genus=unlist(strsplit(species, split=" "))[1],
                host= assoc[Bin.Id],
                type=ifelse(genus %in% get_core() , "Honeybee-associated", "Other"))#,
                #type=ifelse(is.na(host)| host=="Apis mellifera", type, "Other"))  

gen_to_genome<- setNames( clust$genus,clust$Bin.Id)
gen_to_id<- setNames(clust$id,clust$Bin.Id)

# Nr. of bOTUs
length(unique(clust$secondary_cluster))

# Nr. of bee-associated bOTUs
length(unique(clust$secondary_cluster[clust$type=="Honeybee-associated"]))

# Nr. of bee-associated MAGs
length(unique(clust$Bin.Id[clust$type=="Honeybee-associated" & clust$isolate=="no"]))

length(unique(clust$Bin.Id[clust$isolate=="no"]))
length(unique(clust$secondary_cluster[clust$isolate=="yes"]))

clust_stat<- clust %>% filter(type=="Honeybee-associated") %>%
  group_by(secondary_cluster) %>%
  reframe(composition=ifelse(length(unique(isolate))==2, "Both", 
          ifelse("no" %in% isolate, "MAG only", "isolate only")),
          id=unique(id),
          host=assoc[id]) %>%
  group_by(composition)%>%
  reframe(n=length(secondary_cluster))

clust_win<- read.csv(clust_win_p, header=T, sep="\t") %>% 
  group_by(species)%>%
  dplyr::mutate(genus=unlist(strsplit(species, split=" "))[1]) 
```

# Supplementary table of MAGS info
```{r}
clust_info<- fread("../../../results/reference_db_filtered/summary_data_tables/clust_filtered.tsv")

clust_all_info<- checkm_filt %>%
  rename("genome"=`Bin Id`)%>%
  mutate(id=gen_to_id[genome],
         genus=gen_to_genome[genome],
         best_genome=get_mtdata(mtdata_path="../../../results/reference_db_filtered/summary_data_tables/clust_filtered.tsv","ref", corr="Bin.Id", genome),
         sample=ifelse(is.na(sample), "isolate genome", sample)
         )%>%
  relocate(genus, id, .after = genome) %>%
  filter(!is.na(id))%>%
  arrange(id)

#write.table(clust_all_info, file.path(tab_dir, "genomes_info.tsv"), quote=F, sep="\t", row.names = F) 
```


# MAGs size vs isolates
```{r}
clust_ginfo<- checkm_filt %>%
  rename("genome"=`Bin Id`)%>%
  mutate(id=gen_to_id[genome],
         genus=gen_to_genome[genome])%>%
  relocate(id,genus, .after = genome) %>%
  filter(genus %in% get_core()) %>%
  group_by(id) %>%
  mutate(composition=ifelse(length(unique(type))==2, "Both",
                            ifelse(c("MAG") %in% unique(type), "MAGs", "isolates"))) %>%
  filter(composition %in% c("Both", "MAGs"), !genus=="Bombella")

clust_all_info<- checkm_filt %>%
  rename("genome"=`Bin Id`)%>%
  mutate(id=gen_to_id[genome],
         genus=gen_to_genome[genome],
         )%>%
  relocate(id,genus, .after = genome) %>%
  group_by(id) %>%
  mutate(composition=ifelse(length(unique(type))==2, "Both",
                            ifelse(c("MAG") %in% unique(type), "MAGs", "isolates")))

#clust_supp_tab<- write.table(clust_ginfo, file.path(tab_dir, "genomes_info.tsv"), quote=F, sep="\t", row.names = F) 

pal=c("isolate"="hotpink4", "MAG"="lightskyblue4")
library(yangR)
mags_size_plt<- ggplot(clust_ginfo, aes(x=Get_labels_linneus(id), y=`Genome size (bp)`))+
  #geom_point(aes( col=type), size=2,  alpha=0.7)+
  geom_beeswarm(aes( col=type), size=4, alpha=0.7)+
  coord_flip()+
  labs(x="")+
  facet_wrap(.~id, ncol = 1, strip.position = "right", scales = "free_y")+
  theme_minimal()+
  scale_color_manual(values=pal)+
  scale_y_continuous(labels = unit_format(unit = "M", scale = 1e-6))+
  theme(panel.grid.minor.x = element_blank(),
        panel.grid.major = element_line(size = 5/.pt),
        strip.background = element_blank(), strip.text = element_blank(),
        axis.text.x = element_text(face="bold", size=40/.pt),
        axis.text.y = element_text(face="bold", size=40/.pt),
        axis.title = element_text(face="bold", size=40/.pt),
        legend.text = element_text(size=40/.pt,face="bold"),
        legend.title = element_blank(),
        legend.position = "top");mags_size_plt

ggsave(paste(fig_dir, "mags_size_plt.png", sep=""), mags_size_plt,units = "mm",
       height = 297, width = 297, dpi = 600)
```

# Derepliction
```{r}
  
genome_info<-  read.csv("../../../results/inStrain/B/data_tables/all_genome_info.tsv", header=T, sep="\t") %>% 
  mutate("id"= get_mtdata(mtdata_path="../../../results/reference_db_filtered/summary_data_tables/clust_filtered_winners.tsv",
                          "id", corr="genome", genome))%>%
  relocate(species, .after = genome) %>%
  relocate(genus, .after = species) %>%
  relocate(id, .after = genus)

genome_info_filt <- genome_info %>% filter(coverage_median>=5 & breadth >= 0.5) %>%
  group_by(sample) %>%
  mutate("actual_length"=length*breadth,
         norm_reads=filtered_read_pair_count/actual_length,
         "frequency"=norm_reads/sum(norm_reads)) %>%
  relocate(frequency, .before = coverage) %>%
  group_by(sample, genus) %>%
  mutate(freq_inG=norm_reads/sum(norm_reads)) 
```

```{r}
ani_tab_p<- "../../../results/reference_db/data_tables/Ndb.csv"
ani_tab<- read.csv(ani_tab_p, header=T) %>%
  mutate("genus_reference"= get_mtdata(mtdata_path=clust_p, "genus", corr="genome", reference),
         "genus_query"= get_mtdata(mtdata_path=clust_p, "genus", corr="genome", querry),
         "id_reference"=get_mtdata(mtdata_path=clust_p, "id", corr="genome", reference),
         "id_query"=get_mtdata(mtdata_path=clust_p, "id", corr="genome", querry),
         "chosen"=get_mtdata(mtdata_path=clust_p, "ref", corr="genome", reference),
         "secondary_cluster"=get_mtdata(mtdata_path=clust_p, "secondary_cluster", corr="genome", reference),
         "secondary_cluster_n"=get_mtdata(mtdata_path=clust_win_p, "secondary_cluster_n", corr="secondary_cluster", secondary_cluster),
         origin=ifelse(grepl("MAG", reference), "MAG", "Isolate")) %>%
  filter(genus_reference %in% get_core(), genus_query %in% get_core(),
         id_query %in% genome_info_filt$id, id_reference %in% genome_info_filt$id)


ani_pivot<- ani_tab%>%
  arrange(genus_reference, primary_cluster, secondary_cluster)%>%
  pivot_wider(names_from = querry, values_from = ani, values_fill = 0) 
  

ani_matrix<- ani_tab%>%
  arrange(genus_reference, primary_cluster, secondary_cluster)%>%
  pivot_wider(id_cols=reference, names_from = querry, values_from = ani, values_fill = 0) %>%
  column_to_rownames("reference") %>%
  as.matrix()
ani_matrix <- ani_matrix[, rownames(ani_matrix)]


```

```{r plot heatmap function}
fancy_heatmap<- function(genus=""){
  # create matrix
  if(genus==""){
    ani_pivot<- ani_tab%>%
      arrange(genus_reference, primary_cluster, id_reference)%>%
      pivot_wider(names_from = querry, values_from = ani, values_fill = 0) 
    
    ani_matrix<- ani_tab%>% 
      arrange(genus_reference, primary_cluster, id_reference)%>%
      pivot_wider(id_cols=reference, names_from = querry, values_from = ani, values_fill = 0) %>%
      column_to_rownames("reference") %>%
      as.matrix()
  }else{
      ani_pivot<- ani_tab%>% filter(genus_reference==genus, genus_query==genus) %>%
      arrange(genus_reference, primary_cluster, id_reference)%>%
      pivot_wider(names_from = querry, values_from = ani, values_fill = 0) 
    
      ani_matrix<- ani_tab%>%filter(genus_reference==genus, genus_query==genus) %>%
      arrange(genus_reference, primary_cluster, id_reference)%>%
      pivot_wider(id_cols=reference, names_from = querry, values_from = ani, values_fill = 0) %>%
      column_to_rownames("reference") %>%
      as.matrix()
    }
  
  ani_matrix <- ani_matrix[, rownames(ani_matrix)]*100
  
  #chose color scheme for heamap
  col_vec = c("seashell2",viridis::magma(5, begin = 0.2, end = 1, direction = -1))
  col_fun<- colorRamp2(c(0.79, 0.8,0.85,0.9,0.95, 1)*100, col_vec)
  
  # define heatmap annotation
  
  ## Genus
  genus_ann<- setNames(ani_pivot$reference, ani_pivot$genus_reference)
  genus_ann<- genus_ann[!duplicated(genus_ann)]
  genus_pal<- get_core_cols()[names(genus_ann)]
  genus_ann<- names(genus_ann)
  
  ## Id
  id_ann<- setNames(ani_pivot$reference, ani_pivot$id_reference)
  id_ann<- id_ann[!duplicated(id_ann)]
  id_pal<- get_species_cols(names(id_ann))[names(id_ann)]
  id_ann<- names(id_ann)
  id_ann<- Get_labels_linneus(id_ann)
  names(id_pal)<- Get_labels_linneus(names(id_pal))
  
  ## isolate vs MAG
  org_pal<- c("MAG"="lightpink1", "Isolate"= "steelblue3")
  org_ann<- setNames(ani_pivot$reference, ani_pivot$origin)
  org_ann<- org_ann[!duplicated(org_ann)]
  org_pal<- org_pal[names(org_ann)]
  org_ann<- names(org_ann)
  
  #chosen isolates
  ch_ann<- setNames(ani_pivot$reference, ani_pivot$chosen)
  ch_ann<- ch_ann[!duplicated(ch_ann)]
  ch_ann<- names(ch_ann)
  
  # get row lables
  get_rw_labs<- function(ids_f, lev){
    id_lev<- ids_f[ids_f==lev]
    n<- length(id_lev)
    lab_pos<- ceiling(n/2)
    
    after_pos<- length((lab_pos+1):n)
    
    if((n/2)>0.5){
      lab_vec<- c(rep("", lab_pos-1), lev, rep("", after_pos))
    }else{
      lab_vec<- lev
    }
    
    
    return(lab_vec)
  }

  id_labs<- unlist(lapply(unique(id_ann), function(x) get_rw_labs(id_ann, x)))
  
  # pull togheter annotations
  if(genus==""){
    ha<- HeatmapAnnotation("Genus" = genus_ann,
                       "Cluster" = id_ann,
                       "Genome Type"= anno_simple(org_ann, col=org_pal, border = TRUE,
                                                  pch=ch_ann, pt_gp = gpar(col = "black")),
                       col= list("Genus"= genus_pal, "Cluster"= id_pal,
                                 "Genome Type"=org_pal),
                       border = TRUE
                       )
  }else{
    ha<- HeatmapAnnotation("Genome Type"= anno_simple(org_ann, col=org_pal, border = TRUE,
                                                  gp = gpar(col = "black"),
                                                  pch=ch_ann, pt_gp = gpar(col = "black", size=7)),
                       col= list("Genome Type"=org_pal),
                       border = TRUE)
    
    row_ha<- rowAnnotation(foo=anno_text(id_labs, rot = 0, just="left",
                                         gp = gpar(fontsize = 10)),
                           "Cluster" = id_ann,
                           col= list("Cluster"= id_pal),
                           show_legend=F,
                           border = T)
  }
  
  #splitting_column
  split<- setNames(ani_pivot$reference, ani_pivot$id_reference)
  split<- split[!duplicated(split)]
  split<-names(split)
  
  f_split<- factor(split, levels = unique(split))
  split <- as.numeric(f_split)

  #plot heatmap
  ht<- Heatmap(ani_matrix, column_title=genus,
               cluster_rows = F, cluster_columns=F,
               top_annotation = ha,
               left_annotation = row_ha,
               col=col_fun,
               show_column_names = F,
               name = "% ANI\n",
               show_row_names = F,
               column_split=split,
               row_split=split,
               row_title = NULL,
               border_gp = gpar(col = "black"))
  
  lgd_org = Legend(title = "Genome Type", at=1:2, legend_gp =gpar(fill = c("steelblue3", "lightpink1")),
    labels = c("Isolate", "MAG"))

  return(draw(ht, annotation_legend_list = list(lgd_org)))
}

bombi_ht<- fancy_heatmap(genus="Bombilactobacillus")
bifido_ht<-fancy_heatmap(genus="Bifidobacterium")
gilli_ht<-fancy_heatmap(genus="Gilliamella")
snod_ht<-fancy_heatmap(genus="Snodgrassella")
lacto_ht<-fancy_heatmap(genus="Lactobacillus")
barto_ht<-fancy_heatmap(genus="Bartonella")
Frischi_ht<- fancy_heatmap(genus="Frischella")
comm_ht<- fancy_heatmap(genus="Commensalibacter")
bom_ht<- fancy_heatmap(genus="Bombella")
```


```{r}
levs<- get_gen_col_abs(clust_win$genus)[[1]]
cols<-get_gen_col_abs(clust_win$genus)[[2]]

ref_db_comp_plt<- clust_win %>% group_by(genus) %>%
  dplyr::summarise(genus_n=length(genus)) %>%
  ggplot(aes(x="", y=genus_n, fill=factor(genus, levels = levs)))+
  geom_bar(stat="identity", width=1) +
  ggtitle("reference database composition")+
  labs(subtitle = paste("number of genomes:" ,nrow(clust_win)))+
  scale_fill_manual(values=cols, name="Genus")+
  theme_void()+
  coord_polar("y", start=0); ref_db_comp_plt
```

```{r, results="asis", out.width = '100%'}
plts_p<-"../../../results/reference_db_filtered/figures/secondary_clusters_dendograms"
plots <- list.files(plts_p)

for(i in plots){
  filename <- file.path(plts_p, i)
  a<- knitr::include_graphics(filename)
  print(a)
}

```



# Mapping to Reference Database
```{r}
mapdata_p<- "../../../results/mapping/all_B_mapstats.tsv"

mapdata<- read.csv(mapdata_p, header=T, sep="\t") %>%
  mutate(perc_map=(mapped/total_reads)*100,
         perc_properly=(properly_paired/total_reads)*100,
         perc_singletons=(singletons/total_reads)*100)
```

```{r refstats analysis}
mapped_toRef_plt<- ggplot(mapdata, aes(x=factor(sample, levels=mixedsort(sample)),y=perc_map))+
  geom_bar(stat="identity", position = position_dodge())+
  labs(x="", y="% mapped")+
  scale_y_continuous(breaks = seq(0, 110, by = 10))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, face="bold"),
        axis.text.y = element_text(face="bold")); mapped_toRef_plt
```

# Save Figures
```{r}
library(svglite)
fig_dir<- "../../../Figures/ReferenceDatabase"
dir.create(fig_dir, showWarnings = FALSE)

ggsave(paste(fig_dir, "/mappedToRef.svg", sep=""),  mapped_toRef_plt, units ="mm", width = 297, height = 210, dpi=300)

ggsave(paste(fig_dir, "/Ref_DB_composition.svg", sep=""),  ref_db_comp_plt, units ="mm", width = 297, height = 210, dpi=300)

ggsave(paste(fig_dir, "/MAGs_filtering.svg", sep=""),  MAGs_filtering_plt, units ="mm", width = 297, height = 210, dpi=300)


ggsave(paste(fig_dir, "/MAGs_GenomeSize.svg", sep=""),  mag_genome_size_plt, units ="mm", width = 297, height = 210, dpi=300)
ggsave(paste(fig_dir, "/MAGs_completeness.svg", sep=""),  mag_completeness_plt, units ="mm", width = 297, height = 210, dpi=300)
ggsave(paste(fig_dir, "/MAGs_contamination.svg", sep=""),  mag_contamination_plt, units ="mm", width = 297, height = 210, dpi=300)
ggsave(paste(fig_dir, "/MAGs_StrainHeterogeneity.svg", sep=""),  mag_SH_plt, units ="mm", width = 297, height = 210, dpi=300)


ggsave(paste(fig_dir, "/recovered_MAGs.svg", sep=""),  recovered_MAGs_plt, units ="mm", width = 297, height = 210, dpi=300)

dev.off()
```

```{r}
ggsave("../../../Figures/230324_DMF_FriSeminar/MAGs_filtering.png",
       MAGs_filtering_plt,
       units = "in",
       dpi = 900,
       width = 11.7,
       height = 8.3)

ggsave("../../../Figures/230324_DMF_FriSeminar/recovered_MAGs_taxonomy.png",
       recovered_MAGs_plt,
       units = "in",
       dpi = 900,
       width = 11.7,
       height = 8.3)

ggsave("../../../Figures/230324_DMF_FriSeminar/MAGs_core_GenomeSize.png",
       mag_core_genome_size_plt,
       units = "in",
       dpi = 900,
       width = 11.7,
       height = 8.3)

ggsave("../../../Figures/230324_DMF_FriSeminar/MappingPerc_RefDB.png",
       mapped_toRef_plt,
       units = "in",
       dpi = 900,
       width = 11.7,
       height = 8.3)



png(file="../../../Figures/230324_DMF_FriSeminar/bifido_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
bifido_ht
dev.off()
png(file="../../../Figures/230324_DMF_FriSeminar/lacto_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
lacto_ht
dev.off()
png(file="../../../Figures/230324_DMF_FriSeminar/gilli_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
gilli_ht
dev.off()
png(file="../../../Figures/230324_DMF_FriSeminar/bombi_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
bombi_ht
dev.off()
png(file="../../../Figures/230324_DMF_FriSeminar/barto_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
barto_ht
dev.off()
png(file="../../../Figures/230324_DMF_FriSeminar/snod_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
snod_ht
dev.off()

png(file="../../../Figures/230324_DMF_FriSeminar/Fri_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
Frischi_ht
dev.off()

png(file="../../../Figures/230324_DMF_FriSeminar/comm_ht.png",
    width = 11.7, height =8.3, units = "in", res=900)
comm_ht
dev.off()



```


