---
title: "MAGs analysis"
author: "Mam Malick Sy Ndiaye"
date: "2022-12-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(dendextend)
library(gtools)
library(ggbeeswarm)
source("../useful_func.R")
```

# Metgenomes assembly

```{r open assembly statistics}
asmbl_stats_p<- "../../../results/assembly/HF_assembly/all_HFassemblies_summary_tab.txt"
asmbl_stats<- fread(asmbl_stats_p)
```

```{r plot contigs filtering}
asmbl_stats2<- asmbl_stats %>% group_by(sample, accepted) %>%
  dplyr::summarise("bp"= sum(length),
                   "total_contigs"=length(contig)) %>%
  dplyr::mutate(sam_type=ifelse(grepl("P", sample), "virome", "bacteriome"),
                sam_name=substring(sample,1, nchar(sample)-1))

ggplot(asmbl_stats2, aes(x=reorder(sam_name,as.numeric(sam_name)),
                         y=bp,
                         fill=factor(accepted, levels = c("yes", "no"))))+
  geom_bar(stat = "identity", position=position_dodge())+
  scale_fill_discrete( name="Passes Filter?")+
  labs(x="")+
  theme_classic()+
  facet_grid(sam_type~., scales = "free_y")


ggplot(asmbl_stats2, aes(x=reorder(sam_name,as.numeric(sam_name)),
                         y=total_contigs,
                         fill=factor(accepted, levels = c("yes", "no"))))+
  geom_bar(stat = "identity", position=position_dodge())+
  scale_fill_discrete( name="Passes Filter?")+
  labs(x="")+
  theme_classic()+
  facet_grid(sam_type~., scales = "free_y")
```


# MAGs binning & Filtering
```{r open and format checkm data}
checkm_alls_p<- "../../../results/MAG_binning/checkm_QC/summary/all_MAGs_stats.tsv"
checkm_filt_p<- "../../../results/MAG_binning/checkm_QC/summary/filtered_MAGs_stats.tsv"

checkm_alls<- read.csv(checkm_alls_p, header=T, sep="\t", check.names = F)
checkm_filt<- read.csv(checkm_filt_p, header=T, sep="\t", check.names = F)

filt_mags<- checkm_filt$`Bin Id`
checkm_alls<- checkm_alls %>%
  dplyr::mutate("good"=ifelse(`Bin Id` %in% filt_mags, "yes", "no"))
```

```{r Plot MAGs filtering}
checkm_alls2<- checkm_alls %>% 
  dplyr::mutate(filtered=ifelse(good=="yes", 1, 0)) %>%
  group_by(sample) %>%
  dplyr::summarise(count_good=sum(filtered),
                   tot_count=length(sample)) %>%
  melt()

sam_levs= unique(paste(sort(as.numeric(gsub("B", "", checkm_alls2$sample))), "B", sep=""))

ggplot(checkm_alls2,aes(x=factor(sample, levels = sam_levs), y=value, 
                          fill=factor(variable, levels = c("tot_count", "count_good")))) +
  geom_bar(stat="identity", position = position_dodge())+
  scale_fill_discrete(name = "", labels = c("Total MAGs", "Filtered MAGs"))+
  labs(x="", y= "number of MAGs")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r MAGs retained  vs. Discarded}
checkm_alls %>% group_by(sample, good) %>%
  dplyr::summarise(count=length(good)) %>%
  ggplot(.,aes(x=good, y=count)) +
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  labs(x="", y= "number of MAGs")+
  theme_classic()
```


# MAGs taxonomic classification
```{r funtion to parse taxonomic information of GTDB-TK}
get_taxa_cols<- function(entry){
  tax_l<- unlist(strsplit(entry, ";"))
  tax_l_ref<- lapply(tax_l, function(x) gsub(".*__", "", x))
  
  tax_l_ref[tax_l_ref==""]<- NA
  
  names(tax_l_ref)<- c("domain", "phylum", "class", "order", "family", "genus", "species")
  tax_l_ref$species<- gsub("__*.","", tax_l_ref$species)
  
  return(tax_l_ref)
}
```

```{r open GTDB-TK output and format it}
gtdb_p<- "../../../results/MAG_binning/gtdbtk_classification/gtdbtk.bac120.summary.tsv"
gtdb_raw<- read.csv(gtdb_p, header=T, sep = "\t")

gtdb<- gtdb_raw %>% group_by(user_genome) %>%
  dplyr::mutate("domain"= get_taxa_cols(classification)$domain,
                "pylum"= get_taxa_cols(classification)$pylum,
                "class"= get_taxa_cols(classification)$class,
                "order"= get_taxa_cols(classification)$order,
                "family"= get_taxa_cols(classification)$family,
                "genus"= get_taxa_cols(classification)$genus,
                "species"= get_taxa_cols(classification)$species)

core=c("Bombilactobacillus" , "Commensalibacter", "Lactobacillus","Bifidobacterium","Gilliamella",  "Frischella", "Snodgrassella",  "Bartonella","Apibacter")
```

```{r merge checkm and GTDB-TK data}
gtdb_QC<- left_join(checkm_filt, gtdb, by=c("Bin Id"="user_genome")) %>%
  group_by(genus) %>%
  dplyr::mutate("n_genus"= length(genus))

gtdb_core<- gtdb_QC %>% filter(genus %in% core)
```

```{r Plot MAGs taxonomical inf per samples}
levs<- get_gen_col_abs(gtdb_QC$genus)[[1]]
cols<-get_gen_col_abs(gtdb_QC$genus)[[2]]

gtdb_comp<- gtdb_QC %>% group_by(sample, genus) %>%
  dplyr::summarise(n=length(genus)) %>%
  group_by(sample) %>%
  dplyr::mutate(n_rel=n/sum(n))


ggplot(gtdb_comp, aes(x=factor(sample, levels = sam_levs), y=n, fill=factor(genus, levels = levs)))+
  geom_histogram(stat = "identity", position = "stack")+
  scale_fill_manual(values=cols, name="Genus")+
  labs(y="# MAGs", x="")+
  scale_y_continuous(breaks=seq(0,15, by=1))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggplot(gtdb_comp, aes(x=factor(sample, levels = sam_levs), y=n_rel, fill=factor(genus, levels = levs)))+
  geom_histogram(stat = "identity", position = "stack")+
  scale_fill_manual(values=cols, name="Genus")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggplot(gtdb_comp, aes(x="", y=n_rel, fill=factor(genus, levels = levs)))+
  geom_bar(stat="identity", width=1) +
  scale_fill_manual(values=cols, name="Genus")+
  theme_void()+
  coord_polar("y", start=0)
```



```{r Plot MAGs genomics stats}
lab<-paste("n=", gtdb_QC$n_genus, sep="")

# Data for average genome size have been taken from
# Bombilactobacillus (i.e., Firm4): Zheng et al., 2020 (Microbiology Society)
# Other genera: Zheng et al., 2019 (PNAS; Dataset_S01) 
avg_size<- c("Bombilactobacillus"=1.82e6, 
             "Lactobacillus"=1.91e6, 
             "Bifidobacterium"=1.98e6, 
             "Gilliamella"=2.71e6, 
             "Frischella"=2.69e6,
             "Snodgrassella"=2.43e6,
             "Bartonella"=2.64e6,
             "Apibacter"=2.5e6)

gtdb_QC$avg_size<- avg_size[gtdb_QC$genus]

ggplot(gtdb_QC, aes(x=genus, y=`Genome size (bp)`))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_point(aes(y=avg_size), shape=95, size=12, col="red")+
  geom_text(aes(genus, 0, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggplot(gtdb_QC, aes(x=genus, y=Completeness))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_text(aes(genus, 0, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggplot(gtdb_QC, aes(x=genus, y=Contamination))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_text(aes(genus, -1, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggplot(gtdb_QC, aes(x=genus, y=`Strain heterogeneity`))+
  geom_boxplot(outlier.shape = NA)+
  geom_point(position = position_jitter(0.2))+
  geom_text(aes(genus, -1, label = lab),vjust = 1, check_overlap = TRUE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
```{r}

ggplot(gtdb_core, aes(x=`Genome size (bp)`, y= Completeness, group=genus))+
  geom_point()+
  geom_smooth(method="lm",formula=y~log(x))+
  theme_classic()+
  facet_wrap(.~genus)

ggplot(gtdb_core, aes(x=Completeness, y= `Strain heterogeneity`, group=genus))+
  geom_point()+
  geom_smooth(method="lm")+
  theme_classic()+
  facet_wrap(.~genus)
```

# Derepliction
```{r}
clust_p<- "../../../results/reference_db_filtered/summary_data_tables/clust_filtered.tsv"
clust_win_p<- "../../../results/reference_db_filtered/summary_data_tables/clust_filtered_winners.tsv"

clust<- read.csv(clust_p, header=T, sep="\t") %>% 
  group_by(species)%>%
  dplyr::mutate(genus=unlist(strsplit(species, split=" "))[1]) 
  

clust_win<- read.csv(clust_win_p, header=T, sep="\t") %>% 
  group_by(species)%>%
  dplyr::mutate(genus=unlist(strsplit(species, split=" "))[1])
```

```{r}
levs<- get_gen_col_abs(clust_win$genus)[[1]]
cols<-get_gen_col_abs(clust_win$genus)[[2]]

clust_win %>% group_by(genus) %>%
  dplyr::summarise(genus_n=length(genus)) %>%
  ggplot(aes(x="", y=genus_n, fill=factor(genus, levels = levs)))+
  geom_bar(stat="identity", width=1) +
  ggtitle("reference database: species composition")+
  scale_fill_manual(values=cols, name="Genus")+
  theme_void()+
  coord_polar("y", start=0)
```
```{r, results="asis", out.width = '100%'}
plts_p<-"../../../results/reference_db_filtered/figures/secondary_clusters_dendograms"
plots <- list.files(plts_p)

for(i in plots){
  filename <- file.path(plts_p, i)
  a<- knitr::include_graphics(filename)
  print(a)
}

```



# Mapping to Reference Database
```{r}
mapdata_p<- "../../../results/mapping/all_B_mapstats.tsv"

mapdata<- read.csv(mapdata_p, header=T, sep="\t") %>%
  mutate(perc_map=(mapped/total_reads)*100,
         perc_properly=(properly_paired/total_reads)*100,
         perc_singletons=(singletons/total_reads)*100)
```

```{r refstats analysis}
ggplot(mapdata, aes(x=factor(sample, levels=mixedsort(sample)),y=perc_map))+
  geom_bar(stat="identity", position = position_dodge())+
  labs(x="", y="% mapped")+
  scale_y_continuous(breaks = seq(0, 110, by = 10))+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```


